pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -313,6 +725,51 @@ func (cb *ClusterBuilder) buildInboundClusterForPortOrUDS(clusterPort int, bind
 	return localCluster
 }
 
+func (cb *ClusterBuilder) setUpstreamProtocol(node *model.Proxy, c *cluster.Cluster, port *model.Port, direction model.TrafficDirection) {
+	// Add use_downstream_protocol for sidecar proxy only if protocol sniffing is enabled.
+	// Since protocol detection is disabled for gateway and use_downstream_protocol is used
+	// under protocol detection for cluster to select upstream connection protocol when
+	// the service port is unnamed. use_downstream_protocol should be disabled for gateway.
+	if node.Type == model.SidecarProxy && ((util.IsProtocolSniffingEnabledForInboundPort(port) && direction == model.TrafficDirectionInbound) ||
+		(util.IsProtocolSniffingEnabledForOutboundPort(port) && direction == model.TrafficDirectionOutbound)) {
+		// Use downstream protocol. If the incoming traffic use HTTP 1.1, the
+		// upstream cluster will use HTTP 1.1, if incoming traffic use HTTP2,
+		// the upstream cluster will use HTTP2.
+		cb.setUseDownstreamProtocol(c)
+	} else if port.Protocol.IsHTTP2() {
+		cb.setH2Options(c)
+	}
+}
+
+// finalizeClusters does any final cluster field marshaling. This should be called
+// at the end before adding the cluster to list of clusters.
+func (cb *ClusterBuilder) finalizeClusters(clusters []*cluster.Cluster) []*cluster.Cluster {
+	// resolve cluster name conflicts. there can be duplicate cluster names if there are conflicting service definitions.
+	// for any clusters that share the same name the first cluster is kept and the others are discarded.
+	have := sets.Set{}
+	out := make([]*cluster.Cluster, 0, len(clusters))
+	for _, cluster := range clusters {
+		if !have.Contains(cluster.Name) {
+			out = append(out, cb.finalize(cluster))
+		} else {
+			cb.push.AddMetric(model.DuplicatedClusters, cluster.Name, cb.proxy.ID,
+				fmt.Sprintf("Duplicate cluster %s found while pushing CDS", cluster.Name))
+		}
+		have.Insert(cluster.Name)
+	}
+	return out
+}
+
+func (cb *ClusterBuilder) finalize(c *cluster.Cluster) *cluster.Cluster {
+	// Marshall Http Protocol options if they exist.
+	if httpOptions := cb.httpProtocolOptions[c.Name]; httpOptions != nil {
```

Comment:
can not understand the intention of storing `httpProtocolOptions` rather than set it directly.

Reply:
It stored in ClusterBuilder because it is set at multiple places bases on multiple conditions - Envoy only accepts marshalled Any so to avoid repeated marshalling/unmarshalling - I am storing it here.